### PR TITLE
VM ingestion

### PR DESCRIPTION
**Pull request by shwstppr**
## Description
Documentation PR: https://github.com/apache/cloudstack-documentation/pull/84
FS Doc: https://cwiki.apache.org/confluence/display/CLOUDSTACK/VM+Ingestion

VM ingestion feature allows CloudStack to discover, on-board, import existing VMs in an infra. The feature currently works only for VMware, with a hypervisor agnostic framework which may be extended for KVM and XenServer in future.
Two new APIs have been added, `listUnmanagedInstances` and `importUnmanagedInstance`.
### listUnmanagedInstances
API will list all unmanaged virtual machine for a give cluster. Optionally, name for an existing unmanaged virtual machine can be given to retrieve VM details.
**API request params -** 
`clusterid`(UUID of cluster)
`name`(instance name)
**Response -**
`clusterid`
`hostid`
`name`
`osdisplayname`
`memory`
`powerstate`
`cpuCoresPerSocket`
`cpunumber`
`cpuspeed`
`disk`
    - `id`
    - `capacity`
    - `controller`
    - `controllerunit`
    - `imagepath`
    - `position`
`nic`
    - `id`
    - `macaddress`
    - `networkname`
    - `vlanid`
    - `pcislot`

### importUnmanagedInstance
API will import an exisitng unmanaged virtual machine into CloudStack for a given cluster and virtual machine name. Service offering for the VM, disk offerings for volumes and networks for NICs of the VM can be mapped. Some optional parameters like projectid, domainid, hostname, details, etc can also be given. Appropritate networks, service offering and disk offerings need to be present before import and cannot be created on the fly during the API call.
**API request params -** 
`clusterid`(UUID of cluster)
`name`(instance name)
`displayname`
`hostname`
`domainid`
`projectid`
`templateid`
`serviceofferingid`
`nicnetworklist`(Map for NIC ID and corresponding Network UUID)
`datadiskofferinglist`(Map for disk ID and corresponding disk offering UUID)
`details`(Map for VM details)
`migrateallowed`(VM and its volumes are allowed to migrate to different host/pool 
**Response -** 
Same response as that of `deployVirtualMachine` API

A python based script (scripts/vm/hypervisor/vmware/discover_networks.py) has been created to aid migrations. It leverages VMware’s pyvmomi library (https://github.com/vmware/pyvmomi) and allows listing all networks for a vCenter host or cluster.

This script can take following arguments,
```
-h, --help show this help message and exit
-s HOST, --host HOST vSphere service to connect to
-o PORT, --port PORT Port to connect on
-u USER, --user USER User name to use
-p PASSWORD, --password PASSWORD Password to use
-c CLUSTER, --cluster CLUSTER Cluster for which discover networks
-S, --disable_ssl_verification Disable ssl host certificate verification
-d, --debug Debug log messages
```

## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## API sample calls:
**List**
```
> list unmanagedinstances clusterid=079a19c2-2710-4ed6-9110-027dd5e0f262 
{
  "count": 2,
  "unmanagedinstance": [
    {
      "clusterid": "079a19c2-2710-4ed6-9110-027dd5e0f262",
      "cpucorespersocket": 1,
      "cpunumber": 1,
      "cpuspeed": 0,
      "disk": [
        {
          "capacity": 2097152,
          "controller": "ide",
          "controllerunit": 0,
          "id": "51-3001",
          "imagepath": "[8aca30d6f7803420973f01cc91c1c011] test-unmanaged/test-unmanaged_64.vmdk",
          "label": "Hard disk 1",
          "position": 1
        },
        {
          "capacity": 5242880,
          "controller": "scsi",
          "controllerunit": 0,
          "id": "51-2000",
          "imagepath": "[8aca30d6f7803420973f01cc91c1c011] test-unmanaged/test-unmanaged_63.vmdk",
          "label": "Hard disk 2",
          "position": 0
        }
      ],
      "hostid": "4ab4147a-7e40-4abe-a0e2-2fac1aa6f0bf",
      "memory": 512,
      "name": "test-unmanaged",
      "nic": [
        {
          "id": "Network adapter 1",
          "macaddress": "02:00:39:7c:00:24",
          "networkname": "cloud.guest.1212.200.1-vSwitch1",
          "vlanid": 1212
        }
      ],
      "osdisplayname": "CentOS 4/5 or later (64-bit)",
      "powerstate": "POWERED_ON"
    }
  ]
}
```
**Import**
```
> import unmanagedinstance name=test-unmanaged templateid=200f3fe9-c845-11e9-b69f-1e003c0107c4 clusterid=079a19c2-2710-4ed6-9110-027dd5e0f262 serviceofferingid=c4c5b9e8-8a4a-422d-bccc-2f1bcc166c59 datadiskofferinglist[0].disk=51-2000 datadiskofferinglist[0].diskOffering=e9eb069f-2946-4e26-8d81-07aebba868d1 displayname=NewVm nicNetworkList[0].nic="Network adapter 1" nicNetworkList[0].network=4c44312b-f009-4d33-b942-989169f3d2fc
{
  "virtualmachine": {
    "account": "admin",
    "affinitygroup": [],
    "cpunumber": 1,
    "cpuspeed": 500,
    "created": "2019-09-05T11:32:04+0000",
    "details": {},
    "diskofferingid": "e9eb069f-2946-4e26-8d81-07aebba868d1",
    "diskofferingname": "Custom",
    "displayname": "NewVm",
    "displayvm": true,
    "domain": "ROOT",
    "domainid": "20088b4d-c845-11e9-b69f-1e003c0107c4",
    "guestosid": "2015cd08-c845-11e9-b69f-1e003c0107c4",
    "haenable": false,
    "hostid": "4ab4147a-7e40-4abe-a0e2-2fac1aa6f0bf",
    "hostname": "10.2.2.144",
    "hypervisor": "VMware",
    "id": "4fda861e-e693-4227-a8aa-1b4ce90d5041",
    "instancename": "test-unmanaged",
    "isdynamicallyscalable": false,
    "memory": 512,
    "name": "test-unmanaged",
    "nic": [
      {
        "broadcasturi": "vlan://1212",
        "extradhcpoption": [],
        "id": "61952f25-7efa-4bc2-aa92-ecd38c122e65",
        "isdefault": true,
        "isolationuri": "vlan://1212",
        "macaddress": "02:00:39:7c:00:24",
        "networkid": "4c44312b-f009-4d33-b942-989169f3d2fc",
        "networkname": "lol",
        "secondaryip": [],
        "traffictype": "Guest",
        "type": "L2"
      }
    ],
    "ostypeid": "2015cd08-c845-11e9-b69f-1e003c0107c4",
    "passwordenabled": false,
    "rootdeviceid": 0,
    "rootdevicetype": "ROOT",
    "securitygroup": [],
    "serviceofferingid": "c4c5b9e8-8a4a-422d-bccc-2f1bcc166c59",
    "serviceofferingname": "Small Instance",
    "state": "Stopped",
    "tags": [],
    "templatedisplaytext": "CentOS 5.3(64-bit) no GUI (vSphere)",
    "templateid": "200f3fe9-c845-11e9-b69f-1e003c0107c4",
    "templatename": "CentOS 5.3(64-bit) no GUI (vSphere)",
    "userid": "4dfc315c-c845-11e9-b69f-1e003c0107c4",
    "username": "admin",
    "zoneid": "db18d224-279d-4c08-aeef-0cb7030d0edb",
    "zonename": "pr3540-t341-vmware-65u2"
  }
}
```

**Usage of list_vcenter_networks script**
```
./discover_networks.py -s 192.168.1.170 -u administrator@vsphere.local -p password -S

{
  "count": 4,
  "networks": [
    {
      "cluster": "p1-c1",
      "host": "10.2.3.184",
      "portgroup": "cloud.guest.409.200.1-vSwitch1",
      "switch": "vSwitch1",
      "virtualmachines": [
        {
          "device": {
            "label": "Network adapter 1",
            "macaddress": "02:00:37:b5:00:06"
          },
          "name": "i-2-8-VM"
        },
        {
          "device": {
            "label": "Network adapter 1",
            "macaddress": "02:00:24:72:00:02"
          },
          "name": "r-4-VM"
        }
      ],
      "vlanid": "409"
    },
    {
      "cluster": "p1-c1",
      "host": "10.2.3.184",
      "portgroup": "cloud.public.7.200.1-vSwitch1",
      "switch": "vSwitch1",
      "virtualmachines": [
        {
          "device": {
            "label": "Network adapter 3",
            "macaddress": "1e:00:6f:00:00:03"
          },
          "name": "r-4-VM"
        }
      ],
      "vlanid": "7"
    },
    {
      "cluster": "p1-c1",
      "host": "10.2.3.184",
      "portgroup": "cloud.private.untagged.0.1-vSwitch0",
      "switch": "vSwitch0",
      "virtualmachines": [
        {
          "device": {
            "label": "Network adapter 1",
            "macaddress": "02:00:13:e1:00:01"
          },
          "name": "s-1-VM"
        },
        {
          "device": {
            "label": "Network adapter 2",
            "macaddress": "1e:00:07:00:00:19"
          },
          "name": "s-1-VM"
        },
        {
          "device": {
            "label": "Network adapter 1",
            "macaddress": "02:00:22:40:00:02"
          },
          "name": "v-2-VM"
        },
        {
          "device": {
            "label": "Network adapter 2",
            "macaddress": "1e:00:a6:00:00:1a"
          },
          "name": "v-2-VM"
        },
        {
          "device": {
            "label": "Network adapter 2",
            "macaddress": "02:00:72:27:00:04"
          },
          "name": "r-4-VM"
        }
      ],
      "vlanid": "0"
    },
    {
      "cluster": "p1-c1",
      "host": "10.2.3.184",
      "portgroup": "cloud.public.7.0.1-vSwitch1",
      "switch": "vSwitch1",
      "virtualmachines": [
        {
          "device": {
            "label": "Network adapter 3",
            "macaddress": "1e:00:41:00:00:01"
          },
          "name": "s-1-VM"
        },
        {
          "device": {
            "label": "Network adapter 3",
            "macaddress": "1e:00:a6:00:00:02"
          },
          "name": "v-2-VM"
        },
        {
          "device": {
            "label": "Network adapter 4",
            "macaddress": "02:00:50:56:00:05"
          },
          "name": "r-4-VM"
        },
        {
          "device": {
            "label": "Network adapter 5",
            "macaddress": "02:00:20:3d:00:06"
          },
          "name": "r-4-VM"
        }
      ],
      "vlanid": "7"
    }
  ]
}
```

## How Has This Been Tested?

- A VMware added to running CloudStack with multiple hosts.
- A new vitrual machine deployed from VMware vCenter with multiple disks, networks in the cluster.
- At this point CloudStack does not show or control newly deployed virtual machine.
- `listUnmanagedInstances` API is executed for the cluster. It shows newly deployed VM (deployed from vCenter)
- Import API, `importUnmanagedInstance`, is executed, with appropriate service offering, network- VM NIC details and custom disk offerings for this unmanaged VM
- On successful completion of import API, UI is verified showing imported VM. `listVirtualMachines` API is called to verify showing new VM and `cloud` database is checked for containing new records in `vm_instance`, `volumes`, `nics` tables.
- Lifecycle tasks were performed on the imported VM - stop, start, reboot, migrate, attach new disk.

### CODE REVIEW

**Comment by rhtyd** [comment]
@shwstppr is there a doc PR or spec for this feature?
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test matrix

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test cesntos7 vmware6.5

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by andrijapanicsb** [comment]
Abhishek, can you please make sure that this PR is in sync with the latest changes in-house, and address the conflicts.

**Comment by borisstoyanov** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by borisstoyanov** [comment]
@blueorangutan package

File: scripts/vm/hypervisor/vmware/discover_networks.py [hunk 1]
```diff
@@ -0,0 +1,272 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply scripts/vm/hypervisor/vmware/discover_networks.py]: This network discovery script is for discovering networks from VMware datacenter. Admin can run this externally and use the output to create networks in CS for smoother import of existing unmanaged VMs. It will be installed in cloudstack-common package and hence not called anywhere in code.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 2]
```diff
@@ -799,16 +822,14 @@ private Answer execute(ResizeVolumeCommand cmd) {
                 _storageProcessor.expandDatastore(hostDatastoreSystem, dsMo);
             }
 
-            if (vdisk.second() != null && !vdisk.second().toLowerCase().startsWith("scsi"))
-            {
-                s_logger.error("Unsupported disk device bus "+ vdisk.second());
-                throw new Exception("Unsupported disk device bus "+ vdisk.second());
+            if (vdisk.second() != null && !vdisk.second().toLowerCase().startsWith("scsi")) {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Not sure what this meant :)

File: engine/schema/src/main/java/com/cloud/storage/VMTemplateVO.java [hunk 3]
```diff
@@ -230,6 +230,13 @@ public static VMTemplateVO createPreHostIso(Long id, String uniqueName, String n
         return template;
     }
 
+    public static VMTemplateVO createSystemIso(Long id, String uniqueName, String name, boolean isPublic,
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply engine/schema/src/main/java/com/cloud/storage/VMTemplateVO.java]: Import of  VM is allowed without using any template. But since CS needs a template or ISO ID for a VM, we need to create a dummy, removed ISO which can be used for import in such cases. Method to create dummy ISO DB record, https://github.com/apache/cloudstack/pull/3606/files#diff-3eaaf9de1645091dff197645123e669cR200

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java [hunk 4]
```diff
@@ -3950,6 +3950,61 @@ public NicVO savePlaceholderNic(final Network network, final String ip4Address,
...
+                vo.setReservationStrategy(ReservationStrategy.Start);
+                vo.setReservationId(UUID.randomUUID().toString());
+                vo.setIsolationUri(network.getBroadcastUri());
+                vo.setDeviceId(deviceId);
+                vo.setDefaultNic(isDefaultNic);
+                vo = _nicDao.persist(vo);
+                return vo;
+            }
+        });
+        updateNic(vo, network.getId(), 1);
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java]: Purpose here to call `updateNic` is to update active NICS for a network after import a new NIC. I didn't realized it is updating NicVO as well. Let me check if this can be refactored.

File: core/src/main/java/com/cloud/agent/api/GetUnmanagedInstancesCommand.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+package com.cloud.agent.api;
+
+import java.util.List;
+
+@LogLevel(LogLevel.Log4jLevel.Trace)
+public class GetUnmanagedInstancesCommand extends Command {
+
+    String instanceName;
+
+    List<String> managedInstancesNames;
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply core/src/main/java/com/cloud/agent/api/GetUnmanagedInstancesCommand.java]: Should I add code comments here? I'm also working on user documentation for the feature.

File: api/src/main/java/org/apache/cloudstack/vm/UnmanagedInstance.java [hunk 6]
```diff
@@ -0,0 +1,333 @@
...
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.apache.cloudstack.vm;
+
+import java.util.List;
+
+public class UnmanagedInstance {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/org/apache/cloudstack/vm/UnmanagedInstance.java]: Can you please suggest a rename?
This is just a convenience class to pass data retrieved from VMware(or other hypervisors) to managementserver. Since nothing is stored in DB I didn't used `VO` in name. There is a different UnmanagedInstanceResponse class added for API. As far as I know we use TO when we pass objects to agents.

File: api/src/main/java/com/cloud/vm/UserVmService.java [hunk 7]
```diff
@@ -513,4 +513,8 @@ UserVm upgradeVirtualMachine(ScaleVMCmd cmd) throws ResourceUnavailableException
 
     void collectVmNetworkStatistics (UserVm userVm);
 
+    UserVm importVM(final DataCenter zone, final Host host, final VirtualMachineTemplate template, final String instanceName, final String displayName, final Account owner, final String userData, final Account caller, final Boolean isDisplayVm, final String keyboard,
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/com/cloud/vm/UserVmService.java]: There is `importUnmanagedInstance(ImportUnmanagedInstanceCmd)` method in https://github.com/apache/cloudstack/pull/3606/files#diff-3eaaf9de1645091dff197645123e669cR1018
This method is added in UserVmService to keep all user VM commit in the same place.
Let me know if there is need for refactoring.

File: server/src/test/java/org/apache/cloudstack/vm/VmImportManagerImplTest.java [hunk 8]
```diff
@@ -0,0 +1,334 @@
+package org.apache.cloudstack.vm;
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/test/java/org/apache/cloudstack/vm/VmImportManagerImplTest.java]: Added license

File: engine/api/src/main/java/org/apache/cloudstack/engine/orchestration/service/VolumeOrchestrationService.java [hunk 9]
```diff
@@ -127,4 +129,24 @@ DiskProfile allocateTemplatedVolume(Type type, String name, DiskOffering offerin
...
+    /**
+     * Imports an existing volume for a VM into database. Useful while ingesting an unmanaged VM.
+     * @param type Type of the volume - ROOT, DATADISK, etc
+     * @param name Name of the volume
+     * @param offering DiskOffering for the volume
+     * @param size DiskOffering for the volume
+     * @param minIops minimum IOPS for the disk, if not passed DiskOffering value will be used
+     * @param maxIops maximum IOPS for the disk, if not passed DiskOffering value will be used
+     * @param vm VirtualMachine this volume is attached to
+     * @param template Template of the VM of the volume
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply engine/api/src/main/java/org/apache/cloudstack/engine/orchestration/service/VolumeOrchestrationService.java]: Import of VM(and its disk) is allowed without using any template. But since CS needs a template or ISO ID for a VM, we need to create a dummy, removed ISO which can be used for import in such cases. Will try to add javadoc for this somewhere.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 10]
```diff
@@ -430,118 +446,120 @@ public Answer executeRequest(Command cmd) {
 
             Class<? extends Command> clz = cmd.getClass();
             if (cmd instanceof NetworkElementCommand) {
-                return _vrResource.executeRequest((NetworkElementCommand)cmd);
+                return _vrResource.executeRequest((NetworkElementCommand) cmd);
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Must be the Ctrl+Alt+L I used in class and IDE profile I received from someone. Will revert if needed.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java [hunk 11]
```diff
@@ -617,18 +637,19 @@ protected void reconfigureProcessorByHandler(EnumMap<VmwareStorageProcessorConfi
...
      * Examine StorageSubSystem command to get full clone flag, if provided
-     * @param cmd command to execute
+     *
+     * @param cmd    command to execute
      * @param params params
      * @return copy of params including new values, if suitable
      */
     protected EnumMap<VmwareStorageProcessorConfigurableFields, Object> examineStorageSubSystemCommandFullCloneFlagForVmware(CopyCommand cmd,
-            EnumMap<VmwareStorageProcessorConfigurableFields, Object> params) {
+                                                                                                                             EnumMap<VmwareStorageProcessorConfigurableFields, Object> params) {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]: Must be the Ctrl+Alt+L I used in class and IDE profile I received from someone. Will revert if needed.

**Comment by shwstppr** [comment]
@DaanHoogland will address logging related and remaing review comments soon.

File: api/src/main/java/org/apache/cloudstack/api/command/admin/vm/ListUnmanagedInstancesCmd.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        description = "Lists unmanaged virtual machines for a given cluster.",
+        responseObject = UnmanagedInstanceResponse.class,
+        responseView = ResponseObject.ResponseView.Full,
+        entityType = {UnmanagedInstance.class},
+        requestHasSensitiveInfo = false,
+        responseHasSensitiveInfo = true,
+        authorized = {RoleType.Admin},
+        since = "4.14.0")
+public class ListUnmanagedInstancesCmd extends BaseListCmd {
+    public static final Logger s_logger = Logger.getLogger(ListUnmanagedInstancesCmd.class.getName());
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/org/apache/cloudstack/api/command/admin/vm/ListUnmanagedInstancesCmd.java]: changed

File: api/src/main/java/org/apache/cloudstack/vm/UnmanagedInstance.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply api/src/main/java/org/apache/cloudstack/vm/UnmanagedInstance.java]: A TO would be apropriate, than.

File: core/src/main/java/com/cloud/agent/api/GetUnmanagedInstancesCommand.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply core/src/main/java/com/cloud/agent/api/GetUnmanagedInstancesCommand.java]: Either javadoc or regular programmer comment is fine, but let's make sure posterity understands.

File: api/src/main/java/com/cloud/vm/UserVmService.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply api/src/main/java/com/cloud/vm/UserVmService.java]: needed no, but desirable yes. This is a service call so it is open to third party extensions.

File: engine/schema/src/main/java/com/cloud/storage/VMTemplateVO.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply engine/schema/src/main/java/com/cloud/storage/VMTemplateVO.java]: ok, unfortunately I am talking functionality again, but... I think the current state (snapshot'd) is the best solution for template for this VM in such cases, unless there is a common template known that several imported VMs can be referring.

File: plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply plugins/hypervisors/vmware/src/main/java/com/cloud/hypervisor/vmware/resource/VmwareResource.java]:
```
indian hill {
	puts the opening bracket/accolade at the end of the line
	{
		as oposed to below the condition or loop statement it belongs to
	}
} and closes below as the symetric block do
```

File: scripts/vm/hypervisor/vmware/discover_networks.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (DaanHoogland) [reply scripts/vm/hypervisor/vmware/discover_networks.py]: ok, maybe I missed a comment or a usage message?

File: server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java [hunk 13]
```diff
@@ -0,0 +1,1189 @@
...
+        }
+        try {
+            if (!serviceOfferingVO.isDynamic()) {
+                UsageEventUtils.publishUsageEvent(EventTypes.EVENT_VM_IMPORT, userVm.getAccountId(), userVm.getDataCenterId(), userVm.getId(), userVm.getHostName(), serviceOfferingVO.getId(), userVm.getTemplateId(),
+                        userVm.getHypervisorType().toString(), VirtualMachine.class.getName(), userVm.getUuid(), userVm.isDisplayVm());
+            } else {
+                UsageEventUtils.publishUsageEvent(EventTypes.EVENT_VM_IMPORT, userVm.getAccountId(), userVm.getAccountId(), userVm.getDataCenterId(), userVm.getHostName(), serviceOfferingVO.getId(), userVm.getTemplateId(),
+                        userVm.getHypervisorType().toString(), VirtualMachine.class.getName(), userVm.getUuid(), userVm.getDetails(), userVm.isDisplayVm());
+            }
+        } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: added

File: server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java [hunk 14]
```diff
@@ -0,0 +1,1189 @@
...
+        resourceLimitService.incrementResourceCount(userVm.getAccountId(), Resource.ResourceType.user_vm, userVm.isDisplayVm());
+        resourceLimitService.incrementResourceCount(userVm.getAccountId(), Resource.ResourceType.cpu, userVm.isDisplayVm(), new Long(serviceOfferingVO.getCpu()));
+        resourceLimitService.incrementResourceCount(userVm.getAccountId(), Resource.ResourceType.memory, userVm.isDisplayVm(), new Long(serviceOfferingVO.getRamSize()));
+        // Save usage event and update resource count for user vm volumes
+        List<VolumeVO> volumes = volumeDao.findByInstance(userVm.getId());
+        for (VolumeVO volume : volumes) {
+            try {
+                UsageEventUtils.publishUsageEvent(EventTypes.EVENT_VOLUME_CREATE, volume.getAccountId(), volume.getDataCenterId(), volume.getId(), volume.getName(), volume.getDiskOfferingId(), null, volume.getSize(),
+                        Volume.class.getName(), volume.getUuid(), volume.isDisplayVolume());
+            } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: added

File: server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java [hunk 15]
```diff
@@ -0,0 +1,1189 @@
...
+            }
+            try {
+                Volume volume = null;
+                if (vm.getState().equals(VirtualMachine.State.Running)) {
+                    volume = volumeManager.liveMigrateVolume(volumeVO, storagePool);
+                } else {
+                    volume = volumeManager.migrateVolume(volumeVO, storagePool);
+                }
+                if (volume == null) {
+                    throw new Exception();
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: fixed

File: server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java [hunk 16]
```diff
@@ -0,0 +1,1189 @@
...
+            return;
+        }
+        VirtualMachineProfile profile = new VirtualMachineProfileImpl(userVm);
+        // Remove all volumes
+        volumeDao.deleteVolumesByInstance(userVm.getId());
+        // Remove all nics
+        try {
+            networkOrchestrationService.release(profile, true);
+        } catch (Exception e) {
+            nicDao.removeNicsForInstance(userVm.getId());
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: fixed

File: server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java [hunk 17]
```diff
@@ -0,0 +1,1189 @@
...
+                            if (path.contains("-")) {
+                                split = path.split("-");
+                                path = split[0];
+                                if (!Strings.isNullOrEmpty(path) && !path.equals("ROOT") && !additionalNameFilter.contains(path)) {
+                                    volumeFileNames.add(path);
+                                }
+                            }
+                        }
+                    }
+                } catch (Exception e) {
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: added

File: engine/schema/src/main/java/com/cloud/storage/VMTemplateVO.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (shwstppr) [reply engine/schema/src/main/java/com/cloud/storage/VMTemplateVO.java]: This dummy ISO will be created only once and will be used for all imported VMs thereafter when no template is passed in the API call (will be common for all imported VMs).

File: api/src/main/java/org/apache/cloudstack/vm/UnmanagedInstance.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/org/apache/cloudstack/vm/UnmanagedInstance.java]: Renamed

File: core/src/main/java/com/cloud/agent/api/GetUnmanagedInstancesCommand.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (shwstppr) [reply core/src/main/java/com/cloud/agent/api/GetUnmanagedInstancesCommand.java]: Added

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (shwstppr) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/NetworkOrchestrator.java]: Refactored, please check once @DaanHoogland

**Comment by shwstppr** [comment]
@DaanHoogland addressed all review comments that needed changes/refactor, please review

@blueorangutan package

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java [hunk 18]
```diff
@@ -1654,14 +1654,16 @@ public DiskProfile importVolume(Type type, String name, DiskOffering offering, L
...
-        } else if (template.getTemplateType().equals(Storage.TemplateType.DATADISK)) {
-            vol.setTemplateId(template.getId());
-        }
 
-        if (type == Type.ROOT) {
-            vol.setTemplateId(template.getId());
+        if (template != null) {
+            if (template.getFormat() == ImageFormat.ISO) {
+                vol.setIsoId(template.getId());
+            } else if (template.getTemplateType().equals(Storage.TemplateType.DATADISK)) {
```
Comment (DaanHoogland) [inline engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java]: my I suggest 
```} else if (Storage.TemplateType.DATADISK.equals(template.getTemplateType())) {``` as a defense against ```template.getTemplateType()``` maybe being ```null```

File: server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java [hunk 19]
```diff
@@ -233,10 +233,10 @@ private UnmanagedInstanceResponse createUnmanagedInstanceResponse(UnmanagedInsta
         response.setMemory(instance.getMemory());
         response.setOperatingSystemId(instance.getOperatingSystemId());
         response.setOperatingSystem(instance.getOperatingSystem());
-        response.setObjectName(UnmanagedInstance.class.getSimpleName().toLowerCase());
+        response.setObjectName("unmanagedinstance");
```
Comment (DaanHoogland) [inline server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: ? why this change? I am confused both about the initial code and the changed code. Seems not important, but still I am interested in naming per sé. What are the considerations?

Comment (shwstppr) [reply server/src/main/java/org/apache/cloudstack/vm/VmImportManagerImpl.java]: @DaanHoogland Because earlier the class name was `UnmanagedInstance` so it could be easily used as object name for API response but since after review I changed the class name to `UnmanagedInstanceTO` I had to explicitly mention object name

File: engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java (hunk 18 above)
Comment (shwstppr) [reply engine/orchestration/src/main/java/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java]: Done

File: api/src/main/java/com/cloud/vm/VmDetailConstants.java [hunk 20]
```diff
@@ -54,4 +54,12 @@
     String SSH_PUBLIC_KEY = "SSH.PublicKey";
     String PASSWORD = "password";
     String ENCRYPTED_PASSWORD = "Encrypted.Password";
+
+    // VM import with nic, disk and custom params for custom compute offering
+    String NIC = "nic";
+    String NETWORK = "network";
+    String IP_ADDRESS = "ipAddress";
```
[the remark this answers is not in the document]
Comment (shwstppr) [reply api/src/main/java/com/cloud/vm/VmDetailConstants.java]: Changed

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by shwstppr** [comment]
@blueorangutan package

**Comment by borisstoyanov** [comment]
@blueorangutan test

**Comment by rhtyd** [comment]
@DaanHoogland can you do a final review? After that squash-merge.

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@rhtyd I will do a final review and merge once travis and the bba agree

**Comment by DaanHoogland** [comment]
@blueorangutan help

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-65u2

**Comment by DaanHoogland** [comment]
@blueorangutan test centos7 vmware-65u2
